instrumentation/src/com/intellij/rt/coverage/instrumentation/NewTracingInstrumenter.java

```diff
@@ -0,0 +1,120 @@
+/*
+ * Copyright 2000-2018 JetBrains s.r.o.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.intellij.rt.coverage.instrumentation;
+
+import com.intellij.rt.coverage.data.LineData;
+import com.intellij.rt.coverage.data.ProjectData;
+import com.intellij.rt.coverage.instrumentation.data.BranchDataContainer;
+import com.intellij.rt.coverage.instrumentation.data.Jump;
+import com.intellij.rt.coverage.instrumentation.data.Switch;
+import com.intellij.rt.coverage.util.LinesUtil;
+import org.jetbrains.coverage.org.objectweb.asm.*;
+
+public class NewTracingInstrumenter extends AbstractTracingInstrumenter {
+  private static final String BRANCH_HITS_FIELD_NAME = "__$branchHits$__";
+  private static final String BRANCH_HITS_FIELD_TYPE = "[I";
+  private static final String BRANCH_HITS_FIELD_INIT_NAME = "__$branchHitsInit$__";
+  private static final String BRANCH_HITS_LOCAL_VARIABLE_NAME = "__$localBranchHits$__";
+
+  private final String myClassNameType;
+  private final ExtraFieldInstrumenter myExtraFieldInstrumenter;
+
+  public NewTracingInstrumenter(ProjectData projectData, ClassVisitor classVisitor, ClassReader cr, String className, boolean shouldCalculateSource) {
+    super(projectData, classVisitor, className, shouldCalculateSource);
+    myExtraFieldInstrumenter = new ExtraFieldTracingInstrumenter(cr, className);
+    myClassNameType = className.replace(".", "/");
+  }
+
+  @Override
+  public MethodVisitor createTouchCounter(final MethodVisitor mv,
+                                          final BranchDataContainer branchData,
+                                          final LineEnumerator enumerator,
+                                          final int access,
+                                          final String name,
+                                          final String desc,
+                                          final String className) {
+    if (!enumerator.hasExecutableLines()) return myExtraFieldInstrumenter.createMethodVisitor(this, mv, mv, name);
```

Comment:
why you need to do anything if there were no executable lines? I'd expect the same logic in the first if as in old instrumenter

Reply:
If an interface has a <clinit> method we cannot skip it in case of no executable lines as we have to call array 'init' there. I can add more specific condition here.